src/main/scala/util/HeterogeneousBag.scala

```diff
@@ -6,12 +6,34 @@ import Chisel._
 import chisel3.Record
 import scala.collection.immutable.ListMap
 
-final case class HeterogeneousBag[T <: Data](elts: Seq[T]) extends Record with collection.IndexedSeq[T] {
+final case class HeterogeneousBag[T <: Data](elts: Seq[T], eltNames: Option[Seq[String]] = None)
```

Comment:
I would probably make `eltNames: Seq[String]] = Nil` and then replace the `eltNames.foreach`/`eltNames.map` below with `if (eltNames.nonEmpty)`. I say this because AFAICT `Some(Nil)` is not a meaningful case that you want to allow, so it clarifies things to just check emptiness.

Reply:
this wasall reverted